pkg/job/job_test.go

```diff
@@ -0,0 +1,482 @@
+package job_test
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/MinhNghiaD/jobworker/api/worker/proto"
+	"github.com/MinhNghiaD/jobworker/pkg/job"
+	"github.com/sirupsen/logrus"
+)
+
+func TestStartJobs(t *testing.T) {
+	logrus.SetOutput(os.Stderr)
+	manager, err := job.NewManager()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer func() {
+		if err := manager.Cleanup(); err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	// checker
+	checkStartCmd := func(cmd string, args []string) error {
+		jobID, err := manager.CreateJob(cmd, args, "test user")
+
+		if err != nil {
+			return err
+		}
+
+		if j, ok := manager.GetJob(jobID); ok {
+			defer j.Stop(false)
+		}
+
+		return nil
+	}
+
+	// TODO: add more test cases and reinforce the behaviour of the command execution.
+	// NOTE: For now, we accept that user can start job with high privilege. The job will be started but the execution will fail.
+	testcases := []struct {
+		name      string
+		cmd       string
+		args      []string
+		expectErr bool
+	}{
+		{
+			"Empty Command",
+			"",
+			[]string{},
+			true,
+		},
+		{
+			"Non exist command",
+			"abc",
+			[]string{},
+			true,
+		},
+		{
+			"Short term",
+			"ls",
+			[]string{"-la"},
+			false,
+		},
+		{
+			"File access",
+			"mkdir",
+			[]string{"/tmp/testdir"},
+			false,
+		},
+		{
+			"Check user",
+			"whoami",
+			[]string{},
+			false,
+		},
+		{
+			"long running",
+			"top",
+			[]string{"-b"},
+			false,
+		},
+		{
+			"High privilege",
+			"apt",
+			[]string{"update"},
+			false,
+		},
+		{
+			"sudo",
+			"sudo",
+			[]string{"apt", "update"},
+			false,
+		},
+		{
+			"bad args",
+			"ls",
+			[]string{"-wrong"},
+			false,
+		},
+	}
+
+	for _, testCase := range testcases {
+		t.Run(testCase.name, func(t *testing.T) {
+			err := checkStartCmd(testCase.cmd, testCase.args)
+
+			if (err != nil) != (testCase.expectErr) {
+				t.Errorf("Test case %s with input %s returns error %v, while %v error is expected", testCase.name, testCase.cmd, err, testCase.expectErr)
+				return
+			}
+		})
+	}
+}
+
+func TestGetJobStatus(t *testing.T) {
+	logrus.SetOutput(os.Stderr)
+	manager, err := job.NewManager()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer func() {
+		if err := manager.Cleanup(); err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	// checker
+	checkStatusJob := func(cmd string, args []string, expectedStatus *proto.ProcessStatus) error {
+		jobID, err := manager.CreateJob(cmd, args, "test user")
+
+		if err != nil {
+			return err
+		}
+
+		time.Sleep(time.Second)
+
+		j, ok := manager.GetJob(jobID)
+
+		if !ok {
+			return fmt.Errorf("Job not found")
+		}
+
+		defer j.Stop(false)
+
+		time.Sleep(time.Second)
+
+		status := j.Status()
+
+		if (status.Status.State != expectedStatus.State) || (status.Status.ExitCode != expectedStatus.ExitCode) {
+			return fmt.Errorf("Status is %v, when expected %v", status, expectedStatus)
+		}
+
+		return err
```

Comment:
nit: there's no need for space between every line. Use your judgement, but at least errors and `ok` can usually be checked on the immediately following line.

Reply:
Yes, sorry. It's a bad habit developed when I code c++ =)